klangbecken.liq

```diff
@@ -56,23 +56,74 @@ ALSA_DEVICE = getenv_fallback("KLANGBECKEN_ALSA_DEVICE", "default")
 # PLAYLISTS                                         #
 # ================================================= #
 
+# calculate waiting time for repeating a track depending on its playlist
+def calc_wait(playlist) =
+  if playlist == "music" then 259200 # 3 days
+  elsif playlist == "classics" then 86400 # 1 day
+  elsif playlist == "jingles" then 3600 # 1 hour
+  else 0 end
+end
+
+# check if track was played recently
+# parses DATA_DIR/index.json and compares with current time
+skipped = ref 0
+def check_next_func(r) = 
+  # parse index.json
+  index = file.contents(path.concat(DATA_DIR, "index.json"))
```

Comment:
I think that `check_next_func` should work. I do have some questions that we should look into.

* Does `file.contents` work in liquidsoap 1.3.2 or would we need to upgrade to the latest 1.3.x? It looks like the CI tests are being run against 1.1.1 so we would at least need to look into fixing CI so the script can be linted against a less ancient version of liquidsoap.
* Are we sure that `file.contents` doesn't block anything else? As far as I understood it, it's use case is usually to parse small amounts of json and I'm not sure if parsing a large file like `index.json` which is currently 1.3M won't have side-effects on real time stuff like playing audio. A quick glance at the code seems to indicate that it will grow in size maybe compounding realtimeness further if the file size has any effects at all.

Reply:
You certainy have a point there. On second thought it's not the best idea to read the whole file from disk every time a new track is about to be selected. It might be better to use `file.watch` to read the file only when it changes.

The required functionality is certainly there in 1.3.2 according to the docs.

> It looks like the CI tests are being run against 1.1.1 so we would at least need to look into fixing CI so the script can be linted against a less ancient version of liquidsoap.

Yes CI nedds to be fixed. I'll look into it.

Reply:
> You certainy have a point there. On second thought it's not the best idea to read the whole file from disk every time a new track is about to be selected. It might be better to use `file.watch` to read the file only when it changes.

I'm not sure if file.watch will make a big difference since the index file gets written after every track anyway.

Reply:
It's not as optimal as I initially thought, but still better than to parse the file every time the function runs. I see the following benefits:
  - The file won't be read while Klangbecken is 'off air'.
  - The file won't be read when `check_next_func` refuses to play a track and has to run a second (or more) time.
  - I expect `file.watch` to execute the function asynchronously (I'm not sure though).

In general, do you see a better fitted solution to tackling #21?

Reply:
Good points, I guess we should look into how async the file.watch is executed and then go with this solution. The only alternative I can think of is ensuring tracks that shouldn't be played as per #21 don't get added to the playlist when it's generated.

Reply:
I played a little with `file.watch` on my computer and found out that:
- The callback is never executed in parallel. The function runs at most once at the same time.
- The playback is not stopping even when the callback "hangs" for a long time. I tried that with a call to `system("sleep 300")`.


Additionally it behaves quite different whether inotify is installed or not. According to [centos-rpm-liquidsoap/liquidsoap.spec](https://github.com/radiorabe/centos-rpm-liquidsoap/blob/master/liquidsoap.spec) that seems to be the case in our Liquidsoap distribution. The reason is probably [file_watcher_inotify.ml](https://github.com/savonet/liquidsoap/blob/master/src/tools/file_watcher_inotify.ml) vs. [file_watcher_mtime.ml](https://github.com/savonet/liquidsoap/blob/master/src/tools/file_watcher_mtime.ml).

When inotify is installed:
- The callback is executed too many times when it should be executed only once (e. g. Klangbecken goes on air). This seems to depend on the speed of the callback (does not happen when adding `system("sleep X")`.
- When the file is modified while the callback is running liquidsoap seems to 'remember' that and call the function again after it finishes.

When inotify is not installed:
- The callback is only executed once per second, even when the file is modified more often.
- Liquidsoap does not execute the callback again when the file was modified while the callback was running.

Reply:
I got this [reply](https://liquidsoapworkspace.slack.com/archives/CJWSG659U/p1590848363298500) in the Liquidsoap Slack channel:
> @spameier Ok thanks. So, internally, I believe that playlist requests are resolved via our asynchronous scheduler. It should be using one of the "blocking" queue, which means that it's okay and we expect tasks in that queue to potentially take a long time to return.

> In this case, it should be okay if your function starts taking more time to return. You can keep an eye on it and make sure to add more blocking queues if you start having more than 2 of these tasks since that is the default number of such queues. The setting for this is:
> ```
> set("scheduler.generic_queues",2)
> ```
I'll go on with the approach in this PR then.